harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,228 @@
+"""
+autodeepspeed.py is the launch layer for DeepSpeedTrial in Determined.
```

Comment:
what's the convention for "auto" in the naming? why not just deepspeed?

Reply:
i just followed what we do for horovod.  i'm not sure whether it's good to have it be the same as deepspeed's launcher cmd `deepspeed` or to show it's a different launcher.

Reply:
> 4 hours of continuous review, I got a first pass on more than half of it I think.
> 
> I'd like to see more effort in deduplicating code from pytorchtrial. I still firmly believe that sublcassing as a mechanism of sharing code is the _wrong_ way to build a system. But things like `_set_data_loaders()` (if you even support fully custom datasets) could be refactored into standalone functions that don't take a `self` attribute. That is a tricky bit of business logic, and it would be nice to only have one copy of it laying around.

I refactored `_deepspeed_trial.py` and moved some shared methods with `_pytorch_trial.py` into `_metric_utils.py`.  I think the other controller methods are either different enough or simple enough to leave on both controllers.  I haven't flowed the refactor to `_pytorch_trial.py` yet.  Let me know if this seems reasonable to you.